### PR TITLE
More lazy functions

### PR DESCRIPTION
**Pull request by sorhawell**
- add missing lazy functions to `pl$...`
 - rename functions_x files to match py-polars
 - deprecate `$list()`
 - add n_unique, approx_n_unique, head, tail

### CODE REVIEW

**Comment by sorhawell** [comment]
remainder not implement plus some priortization 3 higher

```R
3#map
3#apply
3#fold
3#reduce
3#cumfold
3#cumreduce
3#exclude
3#groups
3#format

2#cumsum
2#any
2#all cover bitwise also
2#quantile
2#arange
2#pl$collect_all
2# pl$repeat
2#pl$arg_where

1#spearman_rank_corr
1#pearson_corr
1#corr
1#cov
1#arg_sort_by
1#duration
1#datetime_
1#date_
1#concat_str_
1#pl$from_epoch
1#pl$rolling_cov
1#pl$rolling_corr
```

File: R/functions__lazy.R [hunk 1]
```diff
@@ -253,6 +253,70 @@ pl$last = function(column = NULL) {#-> Expr | Any:
...
+#'   a = c(1, 8, 3),
+#'   b = c(4, 5, 2),
+#'   c = c("foo", "bar", "foo")
+#' )
+#' df$select(pl$head("a"))
+#'
+#' df$select(pl$head("a",2))
+#'
+#' pl$head(df$get_column("a"),2)
+pl$head = function(column = NULL, n = 10) {#-> Expr | Any:
```
Comment (vincentarelbundock) [inline R/functions__lazy.R:275]: If `NULL` is not an acceptable input, then we may want to leave the argument without default and include an `is.missing()` check instead.

Comment (sorhawell) [reply R/functions__lazy.R:275]: oups good catch :)

**Comment by vincentarelbundock** [comment]
> remainder not implement plus some priortization 3 higher

I'm likely have time to look at some of these next week if you are willing to wait. I can open a separate PR.

**Comment by sorhawell** [comment]
> I'm likely have time to look at some of these next week if you are willing to wait. I can open a separate PR.

Cool @vincentarelbundock  :), I made a task list in #204. You can reserve any you want, but it would anyways be unlikely I would get to finish them all by next week.
I will also wrap up this PR  and start a new one.

File: R/functions__lazy.R [hunk 2]
```diff
@@ -232,6 +253,70 @@ pl$last = function(column = NULL) {#-> Expr | Any:
 }
 
 
+#' Get the first `n` rows.
+#' @name pl_head
+#' @param column if dtype is:
+#' - Series: Take head value in `Series`
+#' - str or in: syntactic sugar for `pl.col(..).head()`
```
Comment (etiennebacher) [inline R/functions__lazy.R:260]:
```suggestion
#' - str or int: syntactic sugar for `pl.col(..).head()`
```
is it int?

File: R/functions__lazy.R [hunk 3]
```diff
@@ -232,6 +253,70 @@ pl$last = function(column = NULL) {#-> Expr | Any:
...
+#' @param n number of rows to take, NULL
+#' @keywords Expr_new
+#' @return Expr or head value of input Series
+#' @examples
+#' df = pl$DataFrame(
```
Comment (etiennebacher) [inline R/functions__lazy.R:265]: shouldn't you add a `$lazy()` here since the function is in `functions__lazy.R`?

File: R/functions__lazy.R [hunk 4]
```diff
@@ -232,6 +253,70 @@ pl$last = function(column = NULL) {#-> Expr | Any:
...
+
+#' Get the last `n` rows.
+#' @name pl_tail
+#' @param column if dtype is:
+#' - Series: Take head value in `Series`
+#' - str or in: syntactic sugar for `pl.col(..).head()`
+#' @param n number of rows to take, NULL
+#' @return Expr or tail value of input Series
+#' @examples
+#' df = pl$DataFrame(
```
Comment (etiennebacher) [inline R/functions__lazy.R:297]: same as lazy comment above

File: R/functions__lazy.R [hunk 5]
```diff
@@ -232,6 +253,70 @@ pl$last = function(column = NULL) {#-> Expr | Any:
...
+  ) |>
+    unwrap("in pl$head():")
+}
+
+
+#' Get the last `n` rows.
+#' @name pl_tail
+#' @param column if dtype is:
+#' - Series: Take head value in `Series`
+#' - str or in: syntactic sugar for `pl.col(..).head()`
```
Comment (etiennebacher) [inline R/functions__lazy.R:293]:
```suggestion
#' - str or in: syntactic sugar for `pl.col(..).tail()`
```

File: R/functions__lazy.R [hunk 6]
```diff
@@ -232,6 +253,70 @@ pl$last = function(column = NULL) {#-> Expr | Any:
...
+    ))
+  ) |>
+    unwrap("in pl$head():")
+}
+
+
+#' Get the last `n` rows.
+#' @name pl_tail
+#' @param column if dtype is:
+#' - Series: Take head value in `Series`
```
Comment (etiennebacher) [inline R/functions__lazy.R:292]:
```suggestion
#' - Series: Take tail value in `Series`
```

File: R/functions__lazy.R [hunk 7]
```diff
@@ -232,6 +253,70 @@ pl$last = function(column = NULL) {#-> Expr | Any:
...
+#' - str or in: syntactic sugar for `pl.col(..).head()`
+#' @param n number of rows to take, NULL
```
Comment (etiennebacher) [inline R/functions__lazy.R:294]: is `NULL` accepted?

File: R/functions__lazy.R [hunk 8]
```diff
@@ -321,10 +406,69 @@ pl$median = function(...) { #-> Expr | Any:
   unwrap("in pl$median():")
 }
 
+#' count n unique values
```
Comment (etiennebacher) [inline R/functions__lazy.R:409]:
```suggestion
#' Count `n` unique values
```

File: R/functions__lazy.R [hunk 9]
```diff
@@ -321,10 +406,69 @@ pl$median = function(...) { #-> Expr | Any:
...
+pl$n_unique = function(column) { #-> int or Expr
+  pcase(
+    inherits(column, c("Series","Expr")), result(column$n_unique()),
+    is_string(column), result(pl$col(column)$n_unique()),
+    or_else = Err(paste("arg [column] is neither Series, Expr or String, but", str_string(column)))
+  ) |>
+    unwrap("in pl$n_unique():")
+}
 
+#' Approx count unique values.
```
Comment (etiennebacher) [inline R/functions__lazy.R:441]:
```suggestion
#' Approximate count of unique values.
```

File: R/functions__lazy.R [hunk 10]
```diff
@@ -321,10 +406,69 @@ pl$median = function(...) { #-> Expr | Any:
...
+#' @name pl_approx_unique
+#' @description This is done using the HyperLogLog++ algorithm for cardinality estimation.
```
Comment (etiennebacher) [inline R/functions__lazy.R:443]: Could mention why/when we need an approximate count

File: R/series__series.R [hunk 11]
```diff
@@ -1046,3 +1046,13 @@ Series_expr = method_as_property(function() {
 Series_to_lit = function() {
   pl$lit(self)
 }
+
+#' Series n_unique
+#' @description return count of unique values in Series
```
Comment (etiennebacher) [inline R/series__series.R:1051]:
```suggestion
#' @description Return count of unique values in Series
```

File: R/series__series.R [hunk 12]
```diff
@@ -1046,3 +1046,13 @@ Series_expr = method_as_property(function() {
 Series_to_lit = function() {
   pl$lit(self)
 }
+
+#' Series n_unique
```
Comment (etiennebacher) [inline R/series__series.R:1050]:
```suggestion
#' Count unique values in Series
```

File: R/expr__expr.R [hunk 13]
```diff
@@ -1942,7 +1942,15 @@ Expr_product = "use_extendr_wrapper"
 #' pl$DataFrame(iris)$select(pl$col("Species")$n_unique())
 Expr_n_unique = "use_extendr_wrapper"
 
-
+#'  Approx count unique values
+#' @keywords Expr
+#' @description
+#' This is done using the HyperLogLog++ algorithm for cardinality estimation.
+#' @aliases approx_unique
+#' @return Expr
```
Comment (etiennebacher) [inline R/expr__expr.R:1950]:
```suggestion
#' @return Expr
#' @docType NULL
```

cf bottom of first post here: https://github.com/pola-rs/r-polars/issues/137#issue-1676063640

This needs to be added everywhere the function is just a character `"use_extendr_wrapper"`

File: NEWS.md [hunk 14]
```diff
@@ -1,5 +1,8 @@
 # polars (development version)
 
+## What's changed
+ - lazy functions translated: `pl$implode`, `pl$explode`, `pl$unique`, `pl$approx_unique`, `pl$head`, `pl$tail`. Deprecated `pl$list` (use `pl$implode`) (#196)
```
Comment (etiennebacher) [inline NEWS.md:4]:
```suggestion
 - lazy functions translated: `pl$implode`, `pl$explode`, `pl$unique`, `pl$approx_unique`, `pl$head`, `pl$tail` (#196)
 - `pl$list` is deprecated, use `pl$implode` instead (#196) 
```

**Review by etiennebacher — CHANGES_REQUESTED** [review]

File: R/functions__lazy.R (hunk 10 above)
Comment (sorhawell) [reply R/functions__lazy.R:448]: added example

**Review by etiennebacher — APPROVED** [review]

**Pull request merged**